### PR TITLE
Refactor label re-alignment in NER pipeline and add tests

### PR DESCRIPTION
**Pull request by francescorubbo**
The refactor addresses bugs for corner cases uncovered when testing each
scenario of label re-alignment with or without ignore_subwords.

@elk-cloner I ended up refactoring a good portion of the re-alignment code while debugging the tests.
Please make sure that the new logic is still implementing the desired behavior.
One aspect I'm unsure about is what score should we assign to the subwords, when using the "default" strategy. For now I'm setting it to 0. Another option would be to setting it to the score for the "O" label, not sure how useful it would be, though.

### CODE REVIEW

File: src/transformers/pipelines/token_classification.py [hunk 1]
```diff
@@ -234,56 +234,68 @@ def __call__(self, inputs: Union[str, List[str]], **kwargs):
...
-            if strategy == "default":
+            if strategy == "default" or strategy is True:
                 label_idx = score[0].argmax()
                 label = self.model.config.id2label[label_idx]
                 sub_words[0]["entity"] = label
+                sub_words[0]["score"] = score[0][label_idx]
                 for sub in sub_words[1:]:
                     sub["entity"] = "O"
-                return sub_words
+                    sub["score"] = 0.0  # what score should we assign here?
```
Comment (francescorubbo) [inline src/transformers/pipelines/token_classification.py:249]: Here the score is set to 0, because we are assigning the arbitrary label "O". Alternatively, we could use the corresponding score.

Comment (elk-cloner) [reply src/transformers/pipelines/token_classification.py:249]: I don't know which one is better, set to 0 or leave it as is??! 
but I think if we don't set the score to 0.0 then we should take for example `sub["score"].max()` or something like this here, because `sub["score"]` is a numpy array with scores for each type of entity.

File: src/transformers/pipelines/token_classification.py [hunk 2]
```diff
@@ -234,56 +234,68 @@ def __call__(self, inputs: Union[str, List[str]], **kwargs):
...
-                max_label_idx = np.mean(score, axis=0).argmax()
-                label = self.model.config.id2label[max_label_idx]
+                for idx, sub in enumerate(sub_words):
+                    sub["entity"] = label
+                    sub["score"] = score[idx][max_label_idx].item()
 
-            for idx, sub in enumerate(sub_words):
-                sub["entity"] = label
-                sub["score"] = score[idx][max_label_idx].item()
+            if self.ignore_subwords:
```
Comment (elk-cloner) [inline src/transformers/pipelines/token_classification.py:268]: I think we don't need to check `self.ignore_subwords` here, we just want to set labels for subwords in `set_labels` function. later according to `self.grouped_entities` we'll decide to merge subwords or not. take a look at [this](https://github.com/huggingface/transformers/issues/10263#issuecomment-782187601) as joshdevins said, I also don't get what is the purpose of `self.ignore_subwords` in the first place. what are your reasons to check `self.ignore_subwords` here ?

File: src/transformers/pipelines/token_classification.py [hunk 3]
```diff
@@ -234,56 +234,68 @@ def __call__(self, inputs: Union[str, List[str]], **kwargs):
...
+        if subword_indices.size == 0:
+            adjacent_subwords = []
+        else:
+            # find non-consecutive indices to identify separate clusters of subwords
+            cluster_edges = np.where(np.diff(subword_indices) != 1)[0]
+            # Sets of adjacent subwords indices, e.g.
+            # ['Sir', 'Test', '##y', 'M', '##c', '##T', '##est', 'is', 'test', '##iful']
+            # --> [[2],[4,5,6],[9]]
+            adjacent_subwords = np.split(subword_indices, cluster_edges + 1)  # shift edge by 1
+
```
Comment (elk-cloner) [inline src/transformers/pipelines/token_classification.py:284]: love your solution to find adjacent subwords 👍👍. I knew that we can probably do it with numpy but didn't figure it out

File: src/transformers/pipelines/token_classification.py (hunk 1 above)
Comment (francescorubbo) [reply src/transformers/pipelines/token_classification.py:249]: Yeah, I think taking the max would be confusing, because it wouldn't match the label. How about we set it to -1? So that it's clear it's not a meaningful score?

File: src/transformers/pipelines/token_classification.py (hunk 2 above)
Comment (francescorubbo) [reply src/transformers/pipelines/token_classification.py:268]: The problem is that `grouped_entities` merges _all_ the consecutive tokens with the same label, not just subwords. In my use case, I need to merge subwords without aggregating multiple words in the same entity.
Considering the example from that thread:
```Max Mustermann → Max Must ##erman ##n```
with `grouped_entities=True and ignore_subwords=True` we get a single entity `"Max Mustermann"`
with `ignore_subwords=True and label_re_alignment=True` we can keep the two words as separate entities `["Max","Mustermann"]`.
This is useful when you need to relabel with additional attributes (in this case it could be `FIRST_NAME`, `LAST_NAME`, for example), or more generally perform subsequent tasks at the word level.
